README
NAME
    CPAN::Meta - the distribution metadata for a CPAN dist

VERSION
    version 2.143240

SYNOPSIS
        use v5.10;
        use strict;
        use warnings;
        use CPAN::Meta;
        use Module::Load;

        my $meta = CPAN::Meta->load_file('META.json');

        printf "testing requirements for %s version %s\n",
        $meta->name,
        $meta->version;

        my $prereqs = $meta->effective_prereqs;

        for my $phase ( qw/configure runtime build test/ ) {
            say "Requirements for $phase:";
            my $reqs = $prereqs->requirements_for($phase, "requires");
            for my $module ( sort $reqs->required_modules ) {
                my $status;
                if ( eval { load $module unless $module eq 'perl'; 1 } ) {
                    my $version = $module eq 'perl' ? $] : $module->VERSION;
                    $status = $reqs->accepts_module($module, $version)
                            ? "$version ok" : "$version not ok";
                } else {
                    $status = "missing"
                };
                say "  $module ($status)";
            }
        }

DESCRIPTION
    Software distributions released to the CPAN include a META.json or, for
    older distributions, META.yml, which describes the distribution, its
    contents, and the requirements for building and installing the
    distribution. The data structure stored in the META.json file is
    described in CPAN::Meta::Spec.

    CPAN::Meta provides a simple class to represent this distribution
    metadata (or *distmeta*), along with some helpful methods for
    interrogating that data.

    The documentation below is only for the methods of the CPAN::Meta
    object. For information on the meaning of individual fields, consult the
    spec.

METHODS
  new
      my $meta = CPAN::Meta->new($distmeta_struct, \%options);

    Returns a valid CPAN::Meta object or dies if the supplied metadata hash
    reference fails to validate. Older-format metadata will be up-converted
    to version 2 if they validate against the original stated specification.

    It takes an optional hashref of options. Valid options include:

    *   lazy_validation -- if true, new will attempt to convert the given
        metadata to version 2 before attempting to validate it. This means
        than any fixable errors will be handled by CPAN::Meta::Converter
        before validation. (Note that this might result in invalid optional
        data being silently dropped.) The default is false.

  create
      my $meta = CPAN::Meta->create($distmeta_struct, \%options);

    This is same as "new()", except that "generated_by" and "meta-spec"
    fields will be generated if not provided. This means the metadata
    structure is assumed to otherwise follow the latest CPAN::Meta::Spec.

  load_file
      my $meta = CPAN::Meta->load_file($distmeta_file, \%options);

    Given a pathname to a file containing metadata, this deserializes the
    file according to its file suffix and constructs a new "CPAN::Meta"
    object, just like "new()". It will die if the deserialized version fails
    to validate against its stated specification version.

    It takes the same options as "new()" but "lazy_validation" defaults to
    true.

  load_yaml_string
      my $meta = CPAN::Meta->load_yaml_string($yaml, \%options);

    This method returns a new CPAN::Meta object using the first document in
    the given YAML string. In other respects it is identical to
    "load_file()".

  load_json_string
      my $meta = CPAN::Meta->load_json_string($json, \%options);

    This method returns a new CPAN::Meta object using the structure
    represented by the given JSON string. In other respects it is identical
    to "load_file()".

  load_string
      my $meta = CPAN::Meta->load_string($string, \%options);

    If you don't know if a string contains YAML or JSON, this method will
    use Parse::CPAN::Meta to guess. In other respects it is identical to
    "load_file()".

  save
      $meta->save($distmeta_file, \%options);

    Serializes the object as JSON and writes it to the given file. The only
    valid option is "version", which defaults to '2'. On Perl 5.8.1 or
    later, the file is saved with UTF-8 encoding.

    For "version" 2 (or higher), the filename should end in '.json'.
    JSON::PP is the default JSON backend. Using another JSON backend
    requires JSON 2.5 or later and you must set the $ENV{PERL_JSON_BACKEND}
    to a supported alternate backend like JSON::XS.

    For "version" less than 2, the filename should end in '.yml'.
    CPAN::Meta::Converter is used to generate an older metadata structure,
    which is serialized to YAML. CPAN::Meta::YAML is the default YAML
    backend. You may set the $ENV{PERL_YAML_BACKEND} to a supported
    alternative backend, though this is not recommended due to subtle
    incompatibilities between YAML parsers on CPAN.

  meta_spec_version
    This method returns the version part of the "meta_spec" entry in the
    distmeta structure. It is equivalent to:

      $meta->meta_spec->{version};

  effective_prereqs
      my $prereqs = $meta->effective_prereqs;

      my $prereqs = $meta->effective_prereqs( \@feature_identifiers );

    This method returns a CPAN::Meta::Prereqs object describing all the
    prereqs for the distribution. If an arrayref of feature identifiers is
    given, the prereqs for the identified features are merged together with
    the distribution's core prereqs before the CPAN::Meta::Prereqs object is
    returned.

  should_index_file
      ... if $meta->should_index_file( $filename );

    This method returns true if the given file should be indexed. It decides
    this by checking the "file" and "directory" keys in the "no_index"
    property of the distmeta structure. Note that neither the version format
    nor "release_status" are considered.

    $filename should be given in unix format.

  should_index_package
      ... if $meta->should_index_package( $package );

    This method returns true if the given package should be indexed. It
    decides this by checking the "package" and "namespace" keys in the
    "no_index" property of the distmeta structure. Note that neither the
    version format nor "release_status" are considered.

  features
      my @feature_objects = $meta->features;

    This method returns a list of CPAN::Meta::Feature objects, one for each
    optional feature described by the distribution's metadata.

  feature
      my $feature_object = $meta->feature( $identifier );

    This method returns a CPAN::Meta::Feature object for the optional
    feature with the given identifier. If no feature with that identifier
    exists, an exception will be raised.

  as_struct
      my $copy = $meta->as_struct( \%options );

    This method returns a deep copy of the object's metadata as an unblessed
    hash reference. It takes an optional hashref of options. If the hashref
    contains a "version" argument, the copied metadata will be converted to
    the version of the specification and returned. For example:

      my $old_spec = $meta->as_struct( {version => "1.4"} );

  as_string
      my $string = $meta->as_string( \%options );

    This method returns a serialized copy of the object's metadata as a
    character string. (The strings are not UTF-8 encoded.) It takes an
    optional hashref of options. If the hashref contains a "version"
    argument, the copied metadata will be converted to the version of the
    specification and returned. For example:

      my $string = $meta->as_string( {version => "1.4"} );

    For "version" greater than or equal to 2, the string will be serialized
    as JSON. For "version" less than 2, the string will be serialized as
    YAML. In both cases, the same rules are followed as in the "save()"
    method for choosing a serialization backend.

STRING DATA
    The following methods return a single value, which is the value for the
    corresponding entry in the distmeta structure. Values should be either
    undef or strings.

    *   abstract

    *   description

    *   dynamic_config

    *   generated_by

    *   name

    *   release_status

    *   version

LIST DATA
    These methods return lists of string values, which might be represented
    in the distmeta structure as arrayrefs or scalars:

    *   authors

    *   keywords

    *   licenses

    The "authors" and "licenses" methods may also be called as "author" and
    "license", respectively, to match the field name in the distmeta
    structure.

MAP DATA
    These readers return hashrefs of arbitrary unblessed data structures,
    each described more fully in the specification:

    *   meta_spec

    *   resources

    *   provides

    *   no_index

    *   prereqs

    *   optional_features

CUSTOM DATA
    A list of custom keys are available from the "custom_keys" method and
    particular keys may be retrieved with the "custom" method.

      say $meta->custom($_) for $meta->custom_keys;

    If a custom key refers to a data structure, a deep clone is returned.

BUGS
    Please report any bugs or feature using the CPAN Request Tracker. Bugs
    can be submitted through the web interface at
    <http://rt.cpan.org/Dist/Display.html?Queue=CPAN-Meta>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

SEE ALSO
    *   CPAN::Meta::Converter

    *   CPAN::Meta::Validator

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/Perl-Toolchain-Gang/CPAN-Meta/issues>. You will be
    notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/Perl-Toolchain-Gang/CPAN-Meta>

      git clone https://github.com/Perl-Toolchain-Gang/CPAN-Meta.git

AUTHORS
    *   David Golden <dagolden@cpan.org>

    *   Ricardo Signes <rjbs@cpan.org>

CONTRIBUTORS
    *   Ansgar Burchardt <ansgar@cpan.org>

    *   Avar Arnfjord Bjarmason <avar@cpan.org>

    *   Christopher J. Madsen <cjm@cpan.org>

    *   Chuck Adams <cja987@gmail.com>

    *   Cory G Watson <gphat@cpan.org>

    *   Damyan Ivanov <dam@cpan.org>

    *   Eric Wilhelm <ewilhelm@cpan.org>

    *   Graham Knop <haarg@haarg.org>

    *   Gregor Hermann <gregoa@debian.org>

    *   Karen Etheridge <ether@cpan.org>

    *   Kenichi Ishigaki <ishigaki@cpan.org>

    *   Ken Williams <kwilliams@cpan.org>

    *   Lars Dieckow <daxim@cpan.org>

    *   Leon Timmermans <leont@cpan.org>

    *   majensen <maj@fortinbras.us>

    *   Mark Fowler <markf@cpan.org>

    *   Matt S Trout <mst@shadowcat.co.uk>

    *   Michael G. Schwern <mschwern@cpan.org>

    *   moznion <moznion@gmail.com>

    *   Olaf Alders <olaf@wundersolutions.com>

    *   Olivier Mengue <dolmen@cpan.org>

    *   Randy Sims <randys@thepierianspring.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by David Golden and Ricardo Signes.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for CPAN-Meta

2.143240  2014-11-20 10:26:30-05:00 America/New_York

  [FIXED]

  - Give correct path in nested merges such as resources

  - Removed strings test that should have been removed when
    CPAN::Meta::Requirements was removed to a separate dist

2.142690  2014-09-26 11:06:34-04:00 America/New_York

  [FIXED]

  - Fixed use of incorrect method in CPAN::Meta::Merge implementation

  [DOCUMENTED]

  - Clarified that no_index is a list of exclusions, and that indexers
    should generally exclude 'inc', 'xt' and 't' as well.
  - CPAN::Meta::History::Meta_1_0 through 1_4 are added as a permanent
    record of 1.x versions of the metaspec

2.142060  2014-07-25 13:30:06-04:00 America/New_York

  [ADDED]

  - CPAN::Meta::Merge is a new class for merging two possibly overlapping
    instances of metadata. It will accept both CPAN::Meta objects and
    (possibly incomplete) hashrefs of metadata.

2.141520  2014-05-31 23:41:13-04:00 America/New_York

  [DOCUMENTED]

  - Clarified use of 'file' for the 'provides' field in the Spec

2.141170  2014-04-27 13:03:37-04:00 America/New_York

  [ADDED]

  - Added ability for CPAN::Meta::Converter to convert metadata fragments
    (incomplete portions of a metadata structure)

  [CHANGED]

  - Optimized internal use of JSON for datastructure cloning

  [FIXED]

  - Removed dependency on List::Util 1.33

  [DOCUMENTED]

  - Clarified language around 'dynamic_config' in the Spec

2.140640  2014-03-05 09:07:05-05:00 America/New_York

  [FIXED]

  - A change to the default meta-spec URL in 2.140630 from search.cpan.org
    to metacpan.org broke Module::Build and ExtUtils::MakeMaker.  This
    reverts that change.  Also added the missing change note to the prior
    version.

2.140630  2014-03-04 15:23:40-05:00 America/New_York

  [CHANGED]

  - Default meta-spec URL points to metacpan.org instead of search.cpan.org

  [ADDED]

  - Added a 'load_string' method that guesses whether the string
    is YAML or JSON

2.133380  2013-12-03 23:01:07 America/New_York

  [FIXED]

  - Improved bad version handling during META conversion

  - When downgrading multiple licenses to version 1.x META formats, if all
    the licenses are open source, the downgraded license will be
    "open_source", not "unknown"

2.132830  2013-10-10 16:04:30 America/New_York

  [ADDED]

  - CPAN::Meta::Prereqs now has a 'merged_requirements' method for
    combining requirements across multiple phases and types

  [FIXED]

  - Invalid 'meta-spec' is no longer a fatal error: instead, it will
    usually be treated as spec version "1.0" (prior to formalization of the
    meta-spec field).  Conversion has some heuristics for guessing a
    version depending on other fields if 'meta-spec' is missing or invalid.

2.132661  2013-09-23 13:27:46 America/New_York

  [FIXED]

  - updated Makefile.PL logic to support PERL_NO_HIGHLANDER

  [PREREQS]

  - Dropped ExtUtils::MakeMaker configure_requires dependency
    to 6.17

2.132660  2013-09-23 06:04:04 America/New_York

  [FIXED]

  - Installation on Perls < 5.12 will uninstall older versions installed
    due to being bundled with ExtUtils::MakeMaker

2.132620  2013-09-19 11:18:33 America/New_York

  [CHANGED]

  - META validation used to allow a scalar value when a list (i.e. array
    reference) was required for a field. This has been tightened and
    validation will now fail if a scalar value is given.  Conversion will
    continue to turn scalars into an array reference as it previously did.

2.132510  2013-09-08 10:17:29 America/New_York

  [FIXED]

  - Fixed incorrectly encoded META.yml

2.132140  2013-08-02 11:54:17 America/New_York

  [DOCUMENTATION]

  - Fixed some typos in CPAN::Meta::Spec

  [OTHER]

  - migrated repository to Perl-Toolchain-Gang organization on Github and
    updated metadata accordingly

2.131560  2013-06-05 15:26:36 America/New_York

  [DOCUMENTATION]

  - Replaced CPAN::Meta SYNOPSIS with code that actually works

2.131490  2013-05-29 14:15:16 America/New_York

  [BUGFIX]

  - Downconversion of custom resources was not dropping the leading "x_".
    Now "x_MailingList" will downconvert correctly to "MailingList".

  [SPEC]

  - Per the Lancaster Consensus, the 'file' subkey of a package listed in
    'provides' must refer to an actual file in the distribution, either the
    .pm file that provides the package or another file (*.PL) that
    generates it

2.130880  2013-03-29 09:51:00 America/New_York

  [BUGFIX]

  - Fix provides 'version' bug introduced in the 2.130870 fix

2.130870  2013-03-28 11:08:50 America/New_York

  [BUGFIX]

  - Don't add 'version' to 'provides' during conversion if one didn't
    exist in the original [Reported by Miyagawa; fixed by David Golden]

2.120921  2012-04-01 15:20:24 Europe/Paris

  [REMOVED]

  - CPAN::Meta::Requirements has been split out into its own distribution
    so it can be used by CPAN.pm without requiring all of CPAN::Meta

2.120920  2012-04-01 11:30:43 Europe/Paris

  [ADDED]

  - CPAN::Meta::Requirements now has a 'requirements_for_module' method
    to retrieve a version requirements string for a specific module
    [Leon Timmermans]

  [OTHER]

  - Parse::CPAN::Meta prerequisite bumped to 1.4403
  - JSON::PP prerequisites bumped to 2.27200
  - CPAN::Meta::YAML prerequisite bumped to 0.008

2.120900  2012-03-30 13:15:15 Europe/Paris

  [BUGFIX]

  - CPAN::Meta::Requirements now treats undef requirements given to
    from_string_hash() as '0' and warns about it; add_string_requirements()
    dies if it does not get a requirements string to parse

2.120630  2012-03-03 14:48:35 EST5EDT

  [BUGFIX]

  - CPAN::Meta::Requirements now ensures that dotted-decimal versions are
    represented in normalized form when stringified.  This fixes a
    regression in META conversion in 2.120620.
    [Reported by Kent Fredric; fixed by David Golden]

2.120620  2012-03-02 12:28:59 EST5EDT

  [ADDED]


lib/CPAN/Meta/History.pm
# vi:tw=72
use 5.006;
use strict;
use warnings;
package CPAN::Meta::History;
our $VERSION = '2.132660'; # VERSION

1;

# ABSTRACT: history of CPAN Meta Spec changes

__END__

=pod

=encoding utf-8

=head1 NAME

CPAN::Meta::History - history of CPAN Meta Spec changes

=head1 VERSION

version 2.132660

=head1 DESCRIPTION

The CPAN Meta Spec has gone through several iterations.  It was
originally written in HTML and later revised into POD (though published
in HTML generated from the POD).  Fields were added, removed or changed,
sometimes by design and sometimes to reflect real-world usage after the
fact.

This document reconstructs the history of the CPAN Meta Spec based on
change logs, repository commit messages and the published HTML files.
In some cases, particularly prior to version 1.2, the exact version
when certain fields were introduced or changed is inconsistent between
sources.  When in doubt, the published HTML files for versions 1.0 to
1.4 as they existed when version 2 was developed are used as the
definitive source.

Starting with version 2, the specification document is part of the
CPAN-Meta distribution and will be published on CPAN as
L<CPAN::Meta::Spec>.

Going forward, specification version numbers will be integers and
decimal portions will correspond to a release date for the CPAN::Meta
library.

=head1 HISTORY

=head2 Version 2

April 2010

=over

=item *

Revised spec examples as perl data structures rather than YAML

=item *

Switched to JSON serialization from YAML

=item *

Specified allowed version number formats

=item *

Replaced 'requires', 'build_requires', 'configure_requires',
'recommends' and 'conflicts' with new 'prereqs' data structure divided
by I<phase> (configure, build, test, runtime, etc.) and I<relationship>
(requires, recommends, suggests, conflicts)

=item *

Added support for 'develop' phase for requirements for maintaining
a list of authoring tools

=item *

Changed 'license' to a list and revised the set of valid licenses

=item *

Made 'dynamic_config' mandatory to reduce confusion

=item *

Changed 'resources' subkey 'repository' to a hash that clarifies
repository type, url for browsing and url for checkout

=item *

Changed 'resources' subkey 'bugtracker' to a hash for either web
or mailto resource

=item *

Changed specification of 'optional_features':

=over

=item *

Added formal specification and usage guide instead of just example

=item *

Changed to use new prereqs data structure instead of individual keys

=back

=item *

Clarified intended use of 'author' as generalized contact list

=item *

Added 'release_status' field to indicate stable, testing or unstable
status to provide hints to indexers

=item *

Added 'description' field for a longer description of the distribution

=item *

Formalized use of "x_" or "X_" for all custom keys not listed in the
official spec

=back

=head2 Version 1.4

June 2008

=over

=item *

Noted explicit support for 'perl' in prerequisites

=item *

Added 'configure_requires' prerequisite type

=item *

Changed 'optional_features'

=over

=item *

Example corrected to show map of maps instead of list of maps
(though descriptive text said 'map' even in v1.3)

=item *

Removed 'requires_packages', 'requires_os' and 'excluded_os'
as valid subkeys

=back

=back

=head2 Version 1.3

November 2006

=over

=item *

Added 'no_index' subkey 'directory' and removed 'dir' to match actual
usage in the wild

=item *

Added a 'repository' subkey to 'resources'

=back

=head2 Version 1.2

August 2005

=over

=item *

Re-wrote and restructured spec in POD syntax

=item *

Changed 'name' to be mandatory

=item *

Changed 'generated_by' to be mandatory

=item *

Changed 'license' to be mandatory

=item *

Added version range specifications for prerequisites

=item *

Added required 'abstract' field

=item *

Added required 'author' field

=item *

Added required 'meta-spec' field to define 'version' (and 'url') of the
CPAN Meta Spec used for metadata

=item *

Added 'provides' field

=item *

Added 'no_index' field and deprecated 'private' field.  'no_index'
subkeys include 'file', 'dir', 'package' and 'namespace'

=item *

Added 'keywords' field

=item *

Added 'resources' field with subkeys 'homepage', 'license', and
'bugtracker'

=item *

Added 'optional_features' field as an alternate under 'recommends'.
Includes 'description', 'requires', 'build_requires', 'conflicts',
'requires_packages', 'requires_os' and 'excluded_os' as valid subkeys

=item *

Removed 'license_uri' field

=back

=head2 Version 1.1

May 2003

=over

=item *

Changed 'version' to be mandatory

=item *

Added 'private' field

=item *

Added 'license_uri' field

=back

=head2 Version 1.0

March 2003

=over

=item *

Original release (in HTML format only)

=item *

Included 'name', 'version', 'license', 'distribution_type', 'requires',
'recommends', 'build_requires', 'conflicts', 'dynamic_config',
'generated_by'

=back

=head1 AUTHORS

=over 4

=item *

David Golden <dagolden@cpan.org>

=item *

Ricardo Signes <rjbs@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by David Golden and Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
